src/parser/Parser.cpp

```diff
@@ -1,122 +1,104 @@
-}
-
-std::unique_ptr<AST::Expr> Parser::parseLiteral() {
-    std::unique_ptr<AST::Expr> statement;
-    switch (m_currentToken.type) {
-        case TokenType::Integer:
-            statement = util::make_unique<AST::IntegerLit>(
-                    AST::IntegerLit(m_currentToken.literal));
-            break;
-        case TokenType::Float:
-            statement = util::make_unique<AST::FloatLit>(
-                    AST::FloatLit(m_currentToken.literal));
-            break;
-        case TokenType::String:
-            statement = util::make_unique<AST::StringLit>(
-                    AST::StringLit(m_currentToken.literal));
-            break;
-        case TokenType::True:
-        case TokenType::False:
-            statement = util::make_unique<AST::BooleanLit>(
-                    AST::BooleanLit(m_currentToken.literal));
-            break;
-        default:
-            // todo : exceptions, details
-            return std::unique_ptr<AST::Expr>(nullptr);
+    if (currTok.type == TokenType::While) {
+        std::unique_ptr<AST::Expr> expr = parseExpr();
+        Token blockBegin = m_tokens.back();
+        if (blockBegin.type != TokenType::LeftBrace) {
+            // TODO syntax error, synchronize
+        }
+        std::unique_ptr<AST::BlockStmt> body = parseBlockStmt();
+        return std::unique_ptr<AST::WhileStmt>(new AST::WhileStmt(std::move(expr), std::move(body)));
+    }
+    if (currTok.type == TokenType::Next) {
+        currTok = m_lexer.getToken();
+        if (currTok.type != TokenType::EOL) {
+            // TODO syntax error, synchronize
+        }
+        return std::unique_ptr<AST::NextStmt>(new AST::NextStmt());
+    }
+    if (currTok.type == TokenType::Break) {
+        currTok = m_lexer.getToken();
+        if (currTok.type != TokenType::EOL) {
+            // TODO syntax error, synchronize
+        }
+        return std::unique_ptr<AST::BreakStmt>(new AST::BreakStmt());
+    }
+    if (currTok.type == TokenType::Let) {
+        // TODO
     }
-    getNextToken();
-    return std::move(statement);
+    if (currTok.type == TokenType::Identifier) {
+        // TODO assign statement
+    }
+    if (currTok.type == TokenType::Print) {
+        std::unique_ptr<AST::Expr> expr = parseExpr();
+        currTok = m_lexer.getToken();
+        if (currTok.type != TokenType::EOL) {
+            // TODO syntax error, synchronize
+        }
+        return std::unique_ptr<AST::PrintStmt>(new AST::PrintStmt(std::move(expr)));
+    }
+    if (currTok.type == TokenType::Err) {
+        std::unique_ptr<AST::Expr> expr = parseExpr();
+        currTok = m_lexer.getToken();
+        if (currTok.type != TokenType::EOL) {
+            // TODO syntax error, synchronize
+        }
+        return std::unique_ptr<AST::ErrStmt>(new AST::ErrStmt(std::move(expr)));
+    }
+    if (currTok.type == TokenType::Scan) {
+        // TODO
+        Token ident = m_lexer.getToken();
+        if (ident.type != TokenType::Identifier) {
```

Comment:
would it be a good idea to add a function where you can say:

```
function errorIf(TokenType, ...TokenTypes) { /* throw syntax error if needed */ }
```

then:

```
errorIf(ident.type, TokenType::Identifier, /* other things this token should not be */)
```

not sure if the `...` syntax is a thing/has an equivalent in cpp but something along these lines might be a useful thing to have for uniform syntax error management

Reply:
What does the `...` syntax do? Is it from Java?

Reply:
In go, it’s basically an array where you can just put stuff in as a bunch of params instead of an actual array

Reply:
Ah, yeah, C++ has [variadic functions](http://en.cppreference.com/w/cpp/language/variadic_arguments) but personally I've never used them.

Reply:
Oh yeah I think that's exactly what I have in mind - would be a convenient way to consistently implement illegal token checks